core/src/main/scala/kafka/server/DelayedFetch.scala

```diff
@@ -71,7 +79,8 @@ class DelayedFetch(delayMs: Long,
    * Case D: The accumulated bytes from all the fetching partitions exceeds the minimum bytes
    * Case E: The partition is in an offline log directory on this broker
    * Case F: This broker is the leader, but the requested epoch is now fenced
-   *
+   * Case G: The high watermark on this broker has changed within a FetchSession, need to propagate to follower (KIP-392)
+   * Case H: The high watermark on this broker has changed during this request, need to propagate to followers (KIP-392)
    * Upon completion, should return whatever data is available for each valid partition
    */
   override def tryComplete(): Boolean = {
```

Comment:
Below we have logic which relies on fully realized LogOffsetMetadata (including segment and position) for the high watermark and LSO. Does this work for the follower since it won't have this information in all cases? 

In general, I think our usage of `LogOffsetMetadata` is unsafe since it's unclear when we can expect the segment position to be present. I think it would be better to only expose the `LogOffsetMetadata` object outside of `Log` when we can rely on its full contents.  We could have two methods in Log for this:

```scala
  def highWatermark: Long = highWatermarkMetadata.messageOffset

  def highWatermarkMetadata: LogOffsetMetadata = {
    if (_highWatermarkMetadata.messageOffsetOnly) {
      _highWatermarkMetadata = lookupOffsetMetadata(_highWatermarkMetadata.messageOffset)
    } 
    _highWatermarkMetadata
  }
```

The code can then be modified to only call `highWatermarkMetadata` when we truly need the segment and position.